generated/advisor/microsoft.advisor/2020-01-01/types.md

```diff
@@ -1,14 +1,23 @@
 # Microsoft.Advisor @ 2020-01-01
 
-## Resource Microsoft.Advisor/configurations@2020-01-01
+## Resource Microsoft.Advisor/configurations@2020-01-01 (WriteOnly)
 * **Valid Scope(s)**: Subscription, ResourceGroup
 ### Properties
 * **apiVersion**: '2020-01-01' (ReadOnly, DeployTimeConstant): The resource api version
 * **id**: string (ReadOnly, DeployTimeConstant): The resource id
 * **name**: 'default' (Required, DeployTimeConstant): The resource name
-* **properties**: [ConfigDataProperties](#configdataproperties): Configuration data properties
+* **properties**: [ConfigDataProperties](#configdataproperties) (WriteOnly): Configuration data properties
```

Comment:
This resource is being flagged correctly as WriteOnly (the [singular configurations endpoint](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/advisor/resource-manager/Microsoft.Advisor/stable/2020-01-01/advisor.json#L143-L203) only defines a `PUT` operation and no `GET` operation), but I don't think that should be filtering down to individual properties. This particular property **is** part of the response the RP returns on resource creation, so it should be something that can be dereferenced off the resource elsewhere in a template. The test that currently determines whether a property is WriteOnly is [a check to see whether the property defined on the resource's `PUT` request is also defined on the resource's `GET` request](https://github.com/Azure/bicep-types-az/blob/1b9d16257cf9bd90d3446b0dd5a1c7f4c02a429c/src/autorest.bicep/src/type-generator.ts#L359-L361), but there are properties that can be specified on a `PUT` that the RP will not echo back in the response, e.g., the certificates provided as part of the creation of an [`Microsoft.Attestation/attestationProviders` resource](https://github.com/Azure/bicep-types-az/blob/9088327d531ead980152cf468401d8a77746a469/generated/attestation/microsoft.attestation/2020-10-01/types.md#attestationservicecreationspecificparams). I believe the latter is what is actually meant by `WriteOnly`, but updating the generator to reflect that meaning is potentially a breaking change (and in any case should happen in a separate PR).

/cc @shenglol

Reply:
My understanding is that if a property is write-only if it can set in PUT requests but is not returned in GET responses. When a resource's property is referenced in a Bicep file, once compiled, it will be transformed into an ARM template expression containing a `reference` function call. To resolve the reference function, the ARM deployment engine will not read the resource PUT response, but it will send a GET request to get the resource body.

Reply:
I think there's a bit more complexity to untangle here, but I don't think this PR is the place to do it. JSON ARM templates will permit reading from a `Microsoft.Advisor/configurations` resource that was created in the same template iff the resource is referenced by name rather than resource ID:

![image](https://user-images.githubusercontent.com/705500/166467291-2fd4159b-aa21-4ad8-848c-3dfe9b696efa.png)
![deref_without_get](https://user-images.githubusercontent.com/705500/166468906-40dc4459-bdb2-4c6b-9c5e-46ad0dc2b2bb.jpg)

When the same template is altered to use a resource ID rather than a resource name, the deployment fails:

![image](https://user-images.githubusercontent.com/705500/166467750-f7459929-df25-401c-9d96-3e9a35eae467.png)
![deref_by_resource_id](https://user-images.githubusercontent.com/705500/166469452-62d581ee-ebc6-4f2f-8f6d-247a60f3aa1f.jpg)

To account for this in the Bicep compiler, properties would need to carry more metadata, and the compiler would need to distinguish between `existing` vs owned resources when evaluating property access. The set of circumstances in which that level of distinction is necessary seems fairly uncommon, since there are only a handful of write-only resources, and I believe most RP APIs return the same structure for both `GET` and `PUT` operations.

Reply:
Good find! Totally agree that this is a complex problem, and it should be out of the scope of this PR. I think I was not entirely correct about how `reference` functions are evaluated. After checking the deployment engine code again, it seems that the deployment engine will always send a GET request if the API version parameter is provided for a `reference` function. Otherwise, it reads the job metadata which contains a resource definition once a resource deployment job is complete. However, whether the resource definition is from a PUT or GET response depends on if the resource type supports asynchronous PUT operations...which we cannot tell from the resource type's Swagger definition.

@jorgecotillo as FYI (and please correct me if I'm wrong).

Reply:
> However, whether the resource definition is from a PUT or GET response depends on if the resource type supports asynchronous PUT operations...which we cannot tell from the resource type's Swagger definition.

Can this be inferred from the presence of a [`x-ms-long-running-operation`](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation) extension? Or is the GET flow only used for resources that use the [`Azure-AsyncOperation` header](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#azure-asyncoperation-resource-format) (which wouldn't be encoded in the Swagger definition)?

Reply:
Ahh yes the `x-ms-long-running-operation` extension can be used (thanks for reminding me!). It needs to be set as long as a resource operation is asynchronous regardless of the use of the `Azure-AsyncOperation` header. The only minor issue might be that there exists RPs who forgot to set the extension for their resource types (e.g., [Blueprint](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/blueprint/resource-manager/Microsoft.Blueprint/preview/2018-11-01-preview/blueprintAssignment.json)...which is our team 😅), and they won't be fixed since that will cause breaking changes in the generated SDKs, but I think we just can treat them as provider errors.